File: PassedProblems/234-palindrome-linked-list.cpp
/*
 * 234-palindrome-linked-list.cpp
 * Given a singly linked list, determine if it is a palindrome.
 * Follow up:
Could you do it in O(n) time and O(1) space?
 *  Created on: Sep 28, 2015
 *      Author: hanzhe
 */


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <iostream>
#include <stack>
using namespace std;
struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
 };
class Solution {
public:

    bool isPalindrome(ListNode* head) {
    	if(head == NULL) return false;
    	if(head->next == NULL)
    		return true;
    	ListNode *f = head, *s = head;
    	while( f->next != NULL && f->next->next != NULL){
    		s = s->next;
    		f = f->next->next;
    	}
    	// reverse second half
    	f = s->next;
//		cout << f->val << endl;
    	while(f != NULL){
    		ListNode * t = f->next;
    		f->next = s;
    		s = f;
    		if(t == NULL)
    			break;
    		f = t;
    	}
//    	cout << "e" << endl;
    	s = head;
    	while(true){
//    		cout << f->val << " " << s->val << endl;
    		if(f->val != s->val)
    			return false;
    		if(f == s && f->next == s)
    			break;
    		f = f->next;
    		s = s->next;
    	}
    	return true;
    }

    bool isPalindrome_O_N_space_cost(ListNode* head) {
    	stack<int> vals;
    	ListNode * t = head;
    	while(t != NULL){
    		vals.push(t->val);
    		t = t->next;
    	}
    	t = head;
    	while(t != NULL){
    		if(t->val != vals.top())
    			return false;
    		t = t->next;
    		vals.pop();
    	}
    	return true;
    }

    void test(){
    	ListNode * l1 = new ListNode(1);
    	ListNode * l2 = new ListNode(2);
//    	l1->next = l2;
    	l2->next = l1;
    	cout << isPalindrome(l2) << endl;
    }
};

File: leetcode/python/234-palindrome-linked-list.py
#!/usr/bin/env python
# encoding: utf-8


"""
Given a singly linked list, determine if it is a palindrome.

Follow up:
Could you do it in O(n) time and O(1) space?
"""


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# 后半部分链表反序，10次过
class Solution(object):
    def isPalindrome(self, head):
        if head == None or head.next == None: return True
        if head.next.next == None: return head.val == head.next.val
        pf, ps = head, head
        while pf != None and pf.next != None and pf.next.next != None:
            pf = pf.next.next
            ps = ps.next
        t, pre = ps.next, ps
        while t != None:
            nt = t.next
            t.next = pre
            t, pre = nt, t
        t = pre
        while t != None and head != None:
            if t.val != head.val: return False
            if t == head or t == head.next: return True
            t, head = t.next, head.next

# 1次过，但是没用O（1)空间
class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        re = []
        t = head
        while t != None:
            re.append(t.val)
            t = t.next
        return re == re[-1::-1]
